File: 卡码/代码随想录/Python/回溯/01-组合/LeetCode_0017.py
#17. 电话号码的字母组合
from typing import *

class Solution:
  def __init__(self):
    self.nums=["","","abc","def","ghi","jkl","mno","pqrs","tuv","wxyz"]
    self.res=[]
  def backtracking(self,digits:str,index:int,tmp:str):
    if index==len(digits):
      self.res.append(tmp)
      return
    digit:int=int(digits[index])
    letter:str=self.nums[digit]
    for ch in letter:
      self.backtracking(digits,index+1,tmp+ch)
  def letterCombinations(self, digits: str) -> List[str]:
    if len(digits)!=0:
      self.backtracking(digits,0,"")
    return self.res

File: 卡码/代码随想录/C++/回溯/01-组合/LeetCode_0017.cpp
//17. 电话号码的字母组合
#include <iostream>
#include <vector>
#include <string>
using namespace std;

class Solution {
	const string nums[10]={"","","abc","def","ghi","jkl","mno","pqrs","tuv","wxyz"};
	vector<string> res;
public:
	void dfs(string digits,int index,const string& str){
		if(index==digits.length()){
			res.push_back(str);
			return;
		}
		int digit=digits[index]-'0';
		string letter=nums[digit];
		for(int i=0;i<letter.length();i++){
			dfs(digits,index+1,str+letter[i]);
		}
	}
	vector<string> letterCombinations(string digits) {
		if(digits.length()!=0) dfs(digits,0,"");
		return res;
	}
};

int main(){
  system("pause");
  return 0;
}
